### PR TITLE
Correctly load matched routes when navigated before preload is finished

### PR DESCRIPTION
**Pull request by freshgiammi**
Ref: https://github.com/TanStack/router/issues/1534
Ref: https://github.com/TanStack/router/issues/1638

### CODE REVIEW

**Comment by freshgiammi** [comment]
There's still an issue here when with context when it gets merged with beforeLoad.

```
export const Route = createFileRoute('/about')({
  beforeLoad: async () => {   
    await new Promise((resolve) => setTimeout(resolve, 5000));
    return { someData: 'hello' };
  },
   loader: async ({context}) => {
    await new Promise((resolve) => setTimeout(resolve, 2000));
    console.log(context.someData)
  },
})
```

Supposing `/about` is the lazy route that takes a long time to resolve, if beforeLoad isn't finished before the navigation event happens, the loader gets invoked with an `undefined` context, breaking the route.

This possibly happens because [execution flow is skipped](https://github.com/TanStack/router/blob/main/packages/react-router/src/router.ts#L1772) when a route already in `beforeLoad` fetching state is re-loaded (because the loader gets invoked regardless of the beforeLoad state, for which we have no pending information on reload)

**Comment by schiller-manuel** [comment]
are you able to reproduce this in a test somehow? this would be awesom.

File: packages/react-router/src/link.tsx [hunk 1]
```diff
@@ -545,6 +545,7 @@ export function useLinkProps<
   options: UseLinkPropsOptions<TRouter, TFrom, TTo, TMaskFrom, TMaskTo>,
 ): React.AnchorHTMLAttributes<HTMLAnchorElement> {
   const router = useRouter()
+  const isClicking = React.useRef(false)
```
Comment (schiller-manuel) [inline packages/react-router/src/link.tsx:548]: why do we need to track if a link was clicked?
what happens if we have two link instances pointing to the same target, this then won't work, right?

i think the solution must be independent from click tracking etc, and also must work when programatically preloading / navigating without clicking on a link

Comment (schiller-manuel) [reply packages/react-router/src/link.tsx:548]: this most likely needs to be fixed inside of `router.ts`

Comment (freshgiammi) [reply packages/react-router/src/link.tsx:548]: This is a different issue than the references (so we could have it in a separate PR), but essentially `onFocus` is also invoked when a button is clicked, which causes the route to get preloaded twice (`onMouseEnter` + `onFocus`).

When focusing a link via keyboard navigation, we obviously skip `onMouseEnter` so it gets correctly preloaded once.
Given that there's no way to distinguish between mouse and keyboard origin for focus events, we can "trap" it by skipping the preload call when a focus event is detected between `onMouseDown` and `onMouseUp`: this makes preload only happen once and ensures that the behaviour is the same for both mouse and keyboard navigation.

**Comment by freshgiammi** [comment]
> are you able to reproduce this in a test somehow? this would be awesom.

The issue I'm still facing in https://github.com/TanStack/router/pull/1876#issuecomment-2202756943 should be easily reproducible.

A test case for this preload issue when lazy components haven't finished fetching yet might be more complicated, but I haven't worked on either yet. Still, it could be something close to what I've done in the [reproduction template](https://stackblitz.com/edit/github-fsj9kg)

File: packages/react-router/src/link.tsx (hunk 1 above)
Comment (schiller-manuel) [reply packages/react-router/src/link.tsx:548]: ok then please put this in a separate PR then

Comment (freshgiammi) [reply packages/react-router/src/link.tsx]: https://github.com/TanStack/router/pull/1885

**Comment by freshgiammi** [comment]
Split https://github.com/TanStack/router/pull/1876#issuecomment-2202756943 into a separate issue: https://github.com/TanStack/router/issues/1886

**Comment by schiller-manuel** [comment]
closing in favor of https://github.com/TanStack/router/pull/1907
I copied the test out of this PR, so thanks for this!

**Pull request closed without merging**